cmd/nerdctl/stats.go

```diff
@@ -0,0 +1,427 @@
+		}
+		cStats.mu.Unlock()
+
+		w := cmd.OutOrStdout()
+		var tmpl *template.Template
+
+		switch format {
+		case "", "table":
+			w = tabwriter.NewWriter(cmd.OutOrStdout(), 10, 1, 3, ' ', 0)
+			fmt.Fprintln(w, "CONTAINER ID\tNAME\tCPU %\tMEM USAGE / LIMIT\tMEM %\tNET I/O\tBLOCK I/O\tPIDS")
+		case "raw":
+			return errors.New("unsupported format: \"raw\"")
+		default:
+			tmpl, err = templates.Parse(format)
+			if err != nil {
+				break
+			}
+		}
+
+		for _, c := range ccstats {
+			rc := statsutil.RenderEntry(&c, noTrunc)
+			if tmpl != nil {
+				var b bytes.Buffer
+				if err := tmpl.Execute(&b, rc); err != nil {
+					break
+				}
+				if _, err = fmt.Fprintf(cmd.OutOrStdout(), b.String()+"\n"); err != nil {
+					break
+				}
+			} else {
+				if _, err := fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n",
+					rc.ID,
+					rc.Name,
+					rc.CPUPerc,
+					rc.MemUsage,
+					rc.MemPerc,
+					rc.NetIO,
+					rc.BlockIO,
+					rc.PIDs,
+				); err != nil {
+					break
+				}
+			}
+		}
+		if f, ok := w.(Flusher); ok {
+			f.Flush()
+		}
+
+		if len(cStats.cs) == 0 && !showAll {
+			break
+		}
+		if noStream {
+			break
+		}
+		select {
+		case err, ok := <-closeChan:
+			if ok {
+				if err != nil {
+					return err
+				}
+			}
+		default:
+			// just skip
+		}
+	}
+
+	return err
+}
+
+func collect(cmd *cobra.Command, s *statsutil.Stats, waitFirst *sync.WaitGroup, id string, noStream bool) {
+
+	logrus.Debugf("collecting stats for %s", s.Container)
+	var (
+		getFirst bool
+		u        = make(chan error, 1)
+	)
+
+	defer func() {
+		// if error happens and we get nothing of stats, release wait group whatever
+		if !getFirst {
+			getFirst = true
+			waitFirst.Done()
+		}
+	}()
+
+	client, ctx, cancel, err := newClient(cmd)
+	if err != nil {
+		s.SetError(err)
+		return
+	}
+	defer cancel()
+	container, err := client.LoadContainer(ctx, id)
+	if err != nil {
+		s.SetError(err)
+		return
+	}
+
+	task, err := container.Task(ctx, nil)
+	if err != nil {
+		s.SetError(err)
+		return
+	}
+
+	clabels, err := container.Labels(ctx)
+	if err != nil {
+		s.SetError(err)
+		return
+	}
+
+	go func() {
+
+		var (
+			previousCgroupCPU     uint64
+			previousCgroupSystem  uint64
+			previousCgroup2CPU    uint64
+			previousCgroup2System uint64
+			previousWindowsCPU    uint64
+			previousWindowsSystem uint64
+			statsEntry            statsutil.StatsEntry
+		)
+
+		for {
+			//sleep to create distant CPU readings
+			time.Sleep(500 * time.Millisecond)
+
+			metric, err := task.Metrics(ctx)
+			if err != nil {
+				u <- err
+				continue
+			}
+			anydata, err := typeurl.UnmarshalAny(metric.Data)
+			if err != nil {
+				u <- err
+				continue
+			}
+			var (
+				data         *v1.Metrics
+				data2        *v2.Metrics
+				windowsStats *wstats.WindowsContainerStatistics
+			)
+			switch v := anydata.(type) {
+			case *v1.Metrics:
+				data = v
+			case *v2.Metrics:
+				data2 = v
+			case *wstats.WindowsContainerStatistics:
+				windowsStats = v
```

Comment:
Part of me thinks it's odd to add Windows logic at the moment, as I don't believe this supports running containers Windows yet.